src/core/internal/array/casting.d

```diff
@@ -25,41 +25,38 @@ Params:
 private void onArrayCastError()(string fromType, size_t fromSize, string toType, size_t toSize) @trusted
 {
     import core.internal.string : unsignedToTempString;
+    import core.memory : pureMalloc;
 
-    const(char)[][9] msgComponents =
-    [
-        "An array of size "
-        , unsignedToTempString(fromSize)
-        , " does not align on an array of size "
-        , unsignedToTempString(toSize)
-        , ", so `"
-        , fromType
-        , "` cannot be cast to `"
-        , toType
-        , "`"
-    ];
-
-    // convert discontiguous `msgComponents` to contiguous string on the stack
+    // convert discontiguous `msgComponents` to contiguous string on the C heap
     enum msgLength = 2048;
-    char[msgLength] msg;
+    // note: never freed!
+    char* msg = cast(char *)pureMalloc(msgLength);
```

Comment:
I don't see any reason to use pureMalloc(). Just use malloc().

Reply:
Not disallowing array casting in pure code? This function is used here https://github.com/dlang/druntime/blob/7d962321244ae1cb03a6c9f25ff85ece83f28277/src/core/internal/array/casting.d#L82

Reply:
Yeah, it wouldn't compile otherwise.